fix grad thrashing of shape analysis

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40939 fix grad thrashing of shape analysis**
* #40938 shape analysis fix for default dtype'

Previously, when we would do shape analysis by running the op with representative inputs, we would always set the grad property to false. This led to a wrong static analysis when we would create differentiable subgraphs, and propagate shapes without also propagating requires_grad, and then uninline them.

Differential Revision: [D22394676](https://our.internmc.facebook.com/intern/diff/D22394676)

Commit:
fix grad thrashing of shape analysis

[ghstack-poisoned]

Krovatkin commented on test/test_jit.py:
```diff
@@ -6802,6 +6802,16 @@ def foo(s: float):
+        torch._C._jit_pass_complete_shape_analysis(foo.graph, (torch.tensor([0.39]),), False)
+
+        # requires_grad property shouldn't be accidentally set by shape analysis
+        self.assertTrue(foo.graph.findNode("aten::sub").output().requiresGrad() is None)
```
couldn't we actually infer in this particular case that `aten::sub` doesn't need `requires_gard`?

eellison replied:
No, we haven’t done grad analysis, we don’t know if it requires grad or not

Krovatkin replied:
oh got it! thanks!

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -421,8 +421,14 @@ class ShapePropagator {
       // its most constrained form.
-      if (stack[i].isTensor())
-        node->outputs()[i]->inferTypeFrom(stack[i].toTensor());
+      auto tensor_type = node->outputs()[i]->type()->cast<TensorType>();
```
Yes, that’s what we’re doing with tensortype::create(stack[i].toTensor())

eellison replied:
Infer type from calls tensortype::create

eellison replied:
Don’t think we can call withRequiresGrad bc that would get rid of grad analysis that has set it to false. I think we just want to preserve whatever our previous grad inference was

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "fix grad thrashing of shape analysis"


Previously, when we would do shape analysis by running the op with representative inputs, we would always set the grad property to false. This led to a wrong static analysis when we would create differentiable subgraphs, and propagate shapes without also propagating requires_grad, and then uninline them.

Differential Revision: [D22394676](https://our.internmc.facebook.com/intern/diff/D22394676)

[ghstack-poisoned]
